rfc/development-cycle.md

```diff
@@ -0,0 +1,175 @@
+
+# Jupyter Development Cycle
+
+The goal of this document is to outline a development cycle for JupyterLab which makes it very easy to do releases. As [documented in this issue](https://github.com/jupyterlab/jupyterlab/issues/8195), it's currently a quite laborious and specialized process. If we can make releases easier to manage, and push any of the hard work around so it's tackled by the authors of the PRs instead of the release managers, we can release more easily and free up core developer time for other matters.
+
+This is meant to support our existing release cycles, not change them, but just provide some standardization and tooling to aid us in keeping to them.
+
+## Assumptions
+
+I have found it helpful to try to first outline the assumptions of our release and development process.
+
+Please give feedback on these first.
+
+You can consider these in addition to the [SemVer 2.0.0](https://semver.org/) spec, having more to do with the relations between the code evolution between multiple versions. They assume we are already following SemVer.
+
+
+**alpha, beta, rc**: Each release goes through some optional number of alpha, beta, and RC release, before the final. For some final release `{x}.{y}.{z}` it can be modeled with this state machine
+
+1. Start by releasing one of these:
+    1. `{x}.{y}.{z}a0`
+    2. `{x}.{y}.{z}b0`
+    3. `{x}.{y}.{z}rc0`
+    4. `{x}.{y}.{z}`
+2. Then the next release can be either an increment of the last number or a release of one of the initial following your
+   release. For example after `{x}.{y}.{z}b10` you could release `{x}.{y}.{z}b11` or `{x}.{y}.{z}rc0` or `{x}.{y}.{z}` (final).
+
+**Previous Ancestor**: The next increment will be the git ancestor of the previous increment.
+
+* Patch: `{x}.{y}.{z}` is an ancestor of `{x}.{y}.{z+1}`.
+* Minor: `{x}.{y}.0` is an ancestor of `{x}.{y+1}.0`.
+* Major: `{x}.0.0` is an ancestor of `{y+1}.0.0`.
+
+**Single Active**: No commits will be queued up for a release until the previous increment was fully release
+(this also falls out of the above assumption). 
+
+* Patch: No commits towards `{x}.{y}.{z}` till `{x}.{y}.{z-1}` final is released.
+* Minor: No commits towards `{x}.{y}.0` till `{x}.{y-1}.0` final is released.
+* Major: No commits towards `{x}.0.0` till `{x-1}.0.0` final is released.
+
+The last two mean you have to totally finish a release on one branch before you start on the next one. For example, it would be illegal to have releases in this order: `1.2.0a0`, `1.3.0a0`, `1.2.0`, then `1.3.0`. However, you could have `1.2.0a0`, `1.1.1a0`, `1.2.0`, then `1.1.1`, because these would be on different branches, `1.x.0` and `1.1.x`.
+
+**JavaScript versions In Sync** We are OK always keeping the JS version bumps in sync. Meaning that if we do a major release of the Python package we also do a major release of all JS versions.
```

Comment:
My major concern with this point is that it will lead to a lot of dead/outdated extensions. I've made a script to profile the state of all extensions with the keyword "jupyterlab-extension" on npm, and got these results:

```
Processed 589 extensions:

Up to date (193)

Outdated (379):
  Support ends at v2.x: 3
  Support ends at v1.x: 173
  Support ends at v0.x: 203
Deprecated (4)

Unclassified (13)
```

(the "Support ends at v2.x" means someone pinned dependency to e.g. "~2.0", so they don't support latest, but support *something* in 2.x).

I'll do a separate PR to share the script, which should ideally be run regularly so we can track the development over time, specially following new releases.

Reply:
Nice work @vidartf!   Those kinds of metrics are important for administrators (and users) looking to decide when to upgrade.

Reply:
@blink1073 Thanks! Any suggestions on were I should put the code?

Reply:
> My major concern with this point is that it will lead to a lot of dead/outdated extensions.

I share this concern. Just want to be clear that this PR tried to make a process our existing way of doing versioning. It changes a few branch names, and would let us release more often, but substantially shouldn't be change the story for extension authors.

I choose to do this, because I thought this would be easier to agree on then having a larger discussion around how to change our release process in general, although I think that would also be useful!

Talking with @kgryte about this, we did bat around some of the ideas from the Node release processes.

Reply:
Also yeah, nice analysis! This is really helpful to have...

Reply:
I just feels like this would formalize something that in my impression is not something "we are OK" with.
```suggestion
```

Reply:
Yeah I agree ! I would love to not contain this part as well. I remember arguing strongly against it in previous meetings, because it forces extension authors to upgrade.

However, it makes it untenable, theoretically, if we want to keep doing backport versioning. Or if we have a practical hack, like bumping new minor version +10 on the next major version, it makes things definitely more confusing.

---

If in turn we could actually version each package independently, which I do think would be ideal for downstream extension authors.

From the PR standpoint, we could conceptually have some way of doing this, having three labels per package. `notebook-extension-minor` or something.

However, I when I started thinking through how we maintain backports for this kind of thing, it starts to get a bit... wild. Haven't totally thought this through, but I started coming to having a branch of basically every combination of packages and their next release.  So maybe like:

`notebook-extension-2.1.x-and-notebook-2.1.x-and...`

IDK I haven't totally figured it out, as I said, but could explore it more to see if anything reasonable could come out of it, though I am doubtful.

Reply:
Couple of good notes from our meeting here as well:

We would hope to see JupyterLab becoming more stable over time, so that means locking down a process, like is documented here, that forces extensions authors to release a new version every time we have a major release, is pretty harmful.

So we *should* figure out a way to not do major version bumps when we don't have to. I don't know how this will work at the moment.

Reply:
One way to actually solve this would be to move out of a monorepo, so that each package could have its own repo and be versioned independently. However, this is a non starter.